### PR TITLE
Optimize: rewrite getbyname_imm and getSRVbyname_imm as wrappers for getby

### PR DESCRIPTION
**Pull request by oknet**

### CODE REVIEW

**Comment by oknet** [comment]
The workflow after PR #4916 
dns-hostdb-pr4916

The workflow with this PR #4961 
dns-hostdb-pr4961

**Comment by SolidWallOfCode** [comment]
Overall I like the cleanup, but I'm quite concerned about the blocking lock. I don't see why
```
(cont->*cb_process_result)(r.get())
```
cannot be bound in to the `HostDBContinuation` if the lock isn't acquired. The continuation can be given the bucket mutex as its mutex, then it need merely try for `lock2` to call the function.

File: iocore/hostdb/HostDB.cc [hunk 1]
```diff
@@ -605,77 +605,106 @@ HostDBContinuation::insert(unsigned int attl)
...
     while (loop) {
       loop = false; // Only loop on explicit set for retry.
       // find the partition lock
-      //
       Ptr<ProxyMutex> bucket_mutex = hostDB.refcountcache->lock_for_key(hash.hash.fold());
       MUTEX_TRY_LOCK(lock2, bucket_mutex, thread);
-
+      if (!lock2.is_locked() && !trylock) {
+        // Refer to: [TS-374](http://issues.apache.org/jira/browse/TS-374)
+        lock2.acquire(thread);
```
[the remark this answers is not in the document]
Comment (oknet) [reply iocore/hostdb/HostDB.cc]: I just rewrote `SCOPED_MUTEX_LOCK(lock, bucket_mutex, thread);`.

**Comment by oknet** [comment]
@SolidWallOfCode 

IMO, the `(cont->*cb_process_result)(r.get())` may be a fast path than `reply_to_cont(cont, r)`.

File: iocore/hostdb/HostDB.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (oknet) [reply iocore/hostdb/HostDB.cc]: Please check the TS-374 commit via ba3b7d023577e33a93cc038ad64f41781633424a.

It solves a performance issue of reverse proxy.

**Comment by SolidWallOfCode** [comment]
Clearly, which is the reason to do a try lock. But if that fails, hard locking and creating blockage from lock contention is the slow path.

**Comment by oknet** [comment]
@SolidWallOfCode 
May I merge this PR first? Because it does not change the logic.
And then open a new issue to discuss the TS-374 (commit ba3b7d0) and `(cont->*cb_process_result)(r.get())`.

The `SCOPED_MUTEX_LOCK` and `(cont->*cb_process_result)(r.get())` are the only differences between the `getbyname_re` and the `getbyname_imm` interfaces.

File: iocore/hostdb/HostDB.cc [hunk 2]
```diff
@@ -605,77 +605,106 @@ HostDBContinuation::insert(unsigned int attl)
...
+  bool trylock          = (cb_process_result == nullptr);
+  bool force_dns        = false;
   EThread *thread       = this_ethread();
   Ptr<ProxyMutex> mutex = thread->mutex;
   ip_text_buffer ipb;
 
+  if (opt.flags & HOSTDB_FORCE_DNS_ALWAYS) {
+    force_dns = true;
+  } else if (opt.flags & HOSTDB_FORCE_DNS_RELOAD) {
+    force_dns = (hostdb_re_dns_on_reload ? true : false);
```
[the remark this answers is not in the document]
Comment (oknet) [reply iocore/hostdb/HostDB.cc]: updated

**Comment by oknet** [comment]
[approve ci autest]

**Comment by vmamidi** [comment]
@oknet I like the code cleanup but i think we might still encounter deadlock like this( #1335 ) even after the cleanup.  #1756 avoids such scenarios but it doesn't prevent anyone from accidentally causing a dead lock similar to ( #1335 ).

**Comment by oknet** [comment]
@vmamidi

I will find a new solution to the TS-374 and replace the commit ba3b7d0.

It will use `MUTEX_TRY_LOCK` completely to avoid the deadlocks.

**Comment by oknet** [comment]
@SolidWallOfCode @vmamidi 
I merged this and open an issue #5043 to discuss the deadlock problem and the solution.

**Pull request merged**